examples/trading/api/indexer/escrow-handler.ts

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+import { SuiEvent } from '@mysten/sui.js/client';
+import { Escrow } from '@prisma/client';
+
+import { prisma } from '../db';
+
+type Optional<T, K extends keyof T> = Pick<Partial<T>, K> & Omit<T, K>;
+
+type EscrowEvent = EscrowCreated | EscrowCancelled | EscrowSwapped;
+
+type EscrowCreated = {
+	sender: string;
+	recipient: string;
+	escrow_id: string;
+	key_id: string;
+};
+
+type EscrowSwapped = {
+	escrow_id: string;
+};
+
+type EscrowCancelled = {
+	escrow_id: string;
+};
+
+/// Handles all events emitted by the `lock` module.
+export const handleEscrowObjects = async (events: SuiEvent[]) => {
+	const updates: Record<
+		string,
+		Optional<Escrow, 'id' | 'keyId' | 'cancelled' | 'sender' | 'recipient' | 'swapped'>
+	> = {};
+
+	for (const event of events) {
```

Comment:
We should maybe be a bit more defensive here -- I know we are only expecting this function to be called by our own code, but just to make sure we don't do something bad when fed bad input, we should:

- Check that the `event.type.startsWith` our package ID, before casting to `EscrowEvent`.
- Explicitly check that the type ends with `EscrowCreated` for the created case.
- Error on an unexpected event, before committing anything to the DB.

Reply:
@amnn there's no public API to this (it's a local process) that writes to the db.

I don't think there's a way for someone to feed it with bad input, as we're strictly defining the events we're listening to (package id + module) when we start the polling loop. Unless the RPC returns invalid data for a valid package/module combo, I don't think that can be tricked!

Reply:
I understand that -- this suggestion is to protect against either an RPC issue, or a bug in the filter that was used to generate the events, rather than a bad actor.

Reply:
Ah I see, that makes sense! thanks